### ISSUE TITLE
ENH: Force real part of Rotation.as_quat() to be positive.

### ISSUE THREAD

**Issue by alexander-soare** [scipy/scipy#17024]
### Describe the solution you'd like.

I think that any return value of `Rotation.as_quat()` should have a non-negative real component. This means we don't have to worry about the fact that quaternions are a double cover on SO(3) and we can always predict what the quaternion components will be given a rotation.

### Describe alternatives you've considered.

Consider a Rotation instance: `r = Rotation.random()`.  And consider the quaternion components I get from it: `q = r.as_quat()`. I can force the real part to be positive without changing the physical significance of the quaternion

```
if q[-1] < 0:
    q = -q
```

**Comment by scottshambaugh** [comment]
Note that with the proposed solution there are still dual points when the scalar component of q is exactly 0, and this extends to the other components. There's probably a more efficient way to write this, but to get a mapping to the same quaternion each time this is the required logic:
```
if ((q[3] < 0)
    or (q[3] == 0 and q[0] < 0)
    or (q[3] == 0 and q[0] == 0 and q[1] < 0)
    or (q[3] == 0 and q[0] == 0 and q[1] == 0 and q[2] < 0)):
    q = -q
```

**Comment by alexander-soare** [comment]
Perhaps more efficient because it avoids rechecking the same thing twice?

```python
for c in np.roll(q):
  if q > 0:
    break
  elif q < 0:
    q = -q
    break
```

**Comment by scottshambaugh** [comment]
@alexander-soare it should all be optimized by the compiler to the same thing, but that might be easier to read so I updated it.

**Comment by alexander-soare** [comment]
@scottshambaugh cool, yeah still getting used to the magic of C compilers.

**Comment by scottshambaugh** [comment]
Psyche, there's [an issue](https://github.com/cython/cython/issues/3160) defining local arrays in cdef's and working around it would be even messier, so I reverted to the old logic.

**Comment by matusvalo** [comment]
> Psyche, there's https://github.com/cython/cython/issues/3160 defining local arrays in cdef's

BTW. the issue was just now fixed and should be released with next cython 3.0 release.
